File: legacy/python/src/usagi/autopilot.py
"""autopilot: 止めるまで走る常駐モード。

現段階では以下を提供:
- start: watch をフォアグラウンド実行（まずは確実に動く形）
- status: stateファイルの存在など簡易表示
- stop: stopファイルを書いて watch ループ側が停止する

次のPRで Discord stop コマンドや PID 管理を強化する。
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path


@dataclass
class AutopilotPaths:
    root: Path

    @property
    def stop_file(self) -> Path:
        return self.root / ".usagi" / "STOP"

    @property
    def state_file(self) -> Path:
        return self.root / ".usagi" / "state.json"


def request_stop(root: Path) -> Path:
    paths = AutopilotPaths(root)
    paths.stop_file.parent.mkdir(parents=True, exist_ok=True)
    paths.stop_file.write_text("stop", encoding="utf-8")
    return paths.stop_file


def clear_stop(root: Path) -> None:
    paths = AutopilotPaths(root)
    if paths.stop_file.exists():
        paths.stop_file.unlink()


def stop_requested(root: Path) -> bool:
    return AutopilotPaths(root).stop_file.exists()

File: internal/autopilot/autopilot.go
package autopilot

import (
	"os"
	"path/filepath"
)

type Paths struct {
	Root string
}

func (p Paths) StopFile() string {
	return filepath.Join(p.Root, ".usagi", "STOP")
}

func (p Paths) StateFile() string {
	return filepath.Join(p.Root, ".usagi", "state.json")
}

func RequestStop(root string) (string, error) {
	p := Paths{Root: root}
	if err := os.MkdirAll(filepath.Dir(p.StopFile()), 0o755); err != nil {
		return "", err
	}
	if err := os.WriteFile(p.StopFile(), []byte("stop"), 0o644); err != nil {
		return "", err
	}
	return p.StopFile(), nil
}

func ClearStop(root string) error {
	p := Paths{Root: root}
	if _, err := os.Stat(p.StopFile()); err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	return os.Remove(p.StopFile())
}

func StopRequested(root string) bool {
	p := Paths{Root: root}
	_, err := os.Stat(p.StopFile())
	return err == nil
}
